src/notebooks/controllers/controllerPreferredService.ts

```diff
@@ -249,12 +249,15 @@ export class ControllerPreferredService implements IControllerPreferredService,
         let preferredConnection: KernelConnectionMetadata | undefined;
         const rankedConnections = await this.kernelRankHelper.rankKernels(
             uri,
+            this.registration.all,
```

Comment:
I don't think rankKernels should internall call the finder API, 
its not possible to rank kernels when we don't already have kernels, hence i don't think rank kernels should call any api to get the kernels, instead it should be given a list of kernsl to rank.
Else the method should be named `getRankedKernels`, and that complicates things.
I think this is simpler now (simpler to adopt the lazy loading of controllers)